wp-content/plugins/tapestry/controller/class.user-controller.php

```diff
@@ -0,0 +1,143 @@
+<?php
+// TODO Change exceptions to using an ERROR class
+/**
+ * Add/update/retrieve User progress
+ * 
+ */
+class TapestryUserController {
+
+    private $userId = NULL;
+
+    public function __construct() {
+        $this->userId = apply_filters('determine_current_user', false); 
+    }
+
+    /** 
+     * Update User's video progress for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @nodeId the current node being viewed
+     * @param type @progressValue is how much the video was viewed, value should be between >= 0 and <= 1
+    */
+    public function updateProgress($postId, $nodeId, $progressValue) {
+        $this->_checkUserAndPostId($this->userId, $postId);
+
+        if ($progressValue !== NULL) {
+            $progressValue = json_decode($progressValue);
+        }
+
+        // Value should be between 0 and 1
+        if ($progressValue < 0 || $progressValue > 1) {
+            throw new Exception('Invalid progress value');
+        }
+
+        $this->_updateUserProgress($this->userId, $postId, $nodeId, $progressValue);
+    }
+
+    /** 
+     * Get User's video progress for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @nodeIdArr is a list of ids currently in the tapestry
+    */
+    public function getProgress($postId, $nodeIdArr) {
+        $this->_isValidTapestryPost($postId);
+        $this->_checkUserAndPostId($this->userId, $postId);
+
+        if ($this->_isJson($nodeIdArr)) {
+            $nodeIdArr = json_decode($nodeIdArr);
+        } else {
+            throw new Exception('Invalid json');
+        }
+
+        return $this->_getUserProgress($this->userId, $postId, $nodeIdArr);
+    }
+
+    /** 
+     * Update User's h5p video setting for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @h5pSettingsData stores volume, playbackRate, quality of h5p video
+    */
+    public function updateH5PSetting($postId, $h5pSettingsData) {
```

Comment:
Please rename this to `updateH5PSettings` since it's multiple settings, not just one setting.

Reply:
renamed to updateH5PSettings